### PR TITLE
Allow to turn off stacktrace bookkeeping

### PR DESCRIPTION
**Pull request by yonromai**
Hi,

The `stack = traceback.extract_stack()[:-1]` instruction makes inserts too slow for my use case.

This PR adds a flag to disable this behavior if needed.

According to some basic testing, it speeds up inserts by ~20x:

Test code:
```python
from sqlitedict import SqliteDict

def insert(outer_stack):
    d = SqliteDict(outer_stack=outer_stack)
    for i in range(10000):
        d["key_{}".format(i)] = "value_{}".format(i)
    d.close()
```
```python
%%timeit
insert(outer_stack=True)
```
=> `2.68 s ± 42.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`

```python
%%timeit
insert(outer_stack=False)
```
=> `121 ms ± 3.04 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)`

### CODE REVIEW

File: sqlitedict.py [hunk 1]
```diff
@@ -109,7 +109,8 @@ class SqliteDict(DictClass):
     VALID_FLAGS = ['c', 'r', 'w', 'n']
 
     def __init__(self, filename=None, tablename='unnamed', flag='c',
-                 autocommit=False, journal_mode="DELETE", encode=encode, decode=decode):
+                 autocommit=False, journal_mode="DELETE", encode=encode, decode=decode,
+                 outer_stack = True):
```
Comment (mpenkov) [inline sqlitedict.py]:
```suggestion
                 outer_stack=True):
```

File: sqlitedict.py [hunk 2]
```diff
@@ -177,7 +181,10 @@ def __init__(self, filename=None, tablename='unnamed', flag='c',
             self.clear()
 
     def _new_conn(self):
-        return SqliteMultithread(self.filename, autocommit=self.autocommit, journal_mode=self.journal_mode)
+        return SqliteMultithread(self.filename,
```
Comment (mpenkov) [inline sqlitedict.py]: Please use a hanging indent. No need to put each parameter on a separate line.

```python
return SqliteMultithread(
    self.filename, autocommit=self.autocommit, journal_mode=self.journal_mode,
    self.outer_stack
)
```
```

File: sqlitedict.py [hunk 3]
```diff
@@ -429,10 +437,14 @@ def run(self):
                         self.log.error(item)
 
                     self.log.error('')  # exception & outer stack w/blank line
-                    self.log.error('Outer stack:')
-                    for item in traceback.format_list(outer_stack):
-                        self.log.error(item)
-                    self.log.error('Exception will be re-raised at next call.')
+                    if not self.outer_stack:
+                        self.log.error('No outer stack to display. '
```
Comment (mpenkov) [inline sqlitedict.py]: This can fit on one line (our hard limit is 110 chars).

**Comment by piskvorky** [comment]
> makes inserts too slow for my use case

Can you post some benchmarks? before / after your PR.

Intuitively, I'd expect debugging features that significantly slow down normal execution to be an opt-in, not opt-out. But if the difference is "slight", it's OK as opt-out for power-users. So the benchmark numbers matter. CC @jquast .

**Comment by yonromai** [comment]
@mpenkov Thanks for the review - Amended the commit based on your comments, PTAL

**Comment by yonromai** [comment]
@piskvorky Although I cannot share my code (not open source), I observed a similar improvement as shown above (about ~23x faster after than before, although I didn't compute the variance in the real use case, as opposed to above). 

FWIW I am evaluating this map compared to a leveldb index and sqlitedict is still ~4x slower than leveldb in my use case.

**Comment by piskvorky** [comment]
Yeah, that wouldn't surprise me. Sqlitedict is aiming to be a simple-to-use wrapper for SQLite, speed is not one of our primary objectives.

**Comment by yonromai** [comment]
@piskvorky LMK if you think I should make this the default behavior

**Comment by piskvorky** [comment]
@yonromai I'm not very familiar with this part of the code. Can you summarize the pros/cons?

What would we (the users, the developers) gain / lose by making it the default?

**Comment by yonromai** [comment]
Pro: Inserts are 23x faster.
Con: If an error occurs, the stack trace will mostly get swallowed, and `'No outer stack to display. Enable it using outer_stack=True'` will be displayed in place of the stacktrace.

**Comment by piskvorky** [comment]
I'm not sure. Leaning slightly toward more intelligible messages… but damn, 23x is a lot! @mpenkov WDYT?

>  'No outer stack to display. Enable it using outer_stack=True' will be displayed

That message is too vague: where should I (as a user) enter this `outer_stack=True`? Can you please make it more specific and actionable.

**Comment by mpenkov** [comment]
@yonromai why did you close this  PR?

**Comment by yonromai** [comment]
@mpenkov I closed the PR because it's been stale and I'm trying to keep my [github.com/pulls](https://github.com/pulls) clean since I use it a lot. 

I kept the fork alive in case you want to cherry pick the changes.

**Comment by mpenkov** [comment]
Are you able to complete the PR? You received feedback from @piskvorky. From what I can see:

1. Document use case (most importantly, source of exceptions)
2. Improve error message

**Comment by yonromai** [comment]
@mpenkov Sorry but I don't have much more bandwidth to dedicate to this (we decided to not use `sqlitedict`).

**Comment by mpenkov** [comment]
OK, that's fair enough. Thanks for the changes in your PR. We'll take it from here.

**Comment by imachug** [comment]
Has there been any progress on this PR? `traceback.extract_stack()` stat(2)'s all Python source files in the traceback--making more syscalls than sqlite itself. No wonder this is so slow.

**Comment by mpenkov** [comment]
@imachug No. Are you interested in fixing this?

**Comment by imachug** [comment]
I thought this PR is more or less ready to merge, except maybe

> That message is too vague: where should I (as a user) enter this outer_stack=True? Can you please make it more specific and actionable.

Do you have any other issues with this?

**Comment by mpenkov** [comment]
No, TBH, I think that's the only thing left. Let me go clean up the error message and we can merge this.

File: sqlitedict.py [hunk 4]
```diff
@@ -140,6 +141,8 @@ def __init__(self, filename=None, tablename='unnamed', flag='c',
         object.
         The default is to use pickle.
 
+        If you disable `outer_stack`, the stacktrace at insert time won't be saved. This
```
Comment (piskvorky) [inline sqlitedict.py:144]: More context / info please. As a user, what is "stacktrace at insert time", when would I need that and why?
